### ISSUE TITLE
Re-structure for pypi

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#3]
In order to distribute PSyclone using the Python Package Index we will need to re-structure things.

**Comment by arporter** [comment]
Although we won't actually re-structure until 1.4 has been released I have made a start on the imaginatively-named branch `pypi_restructure`.

**Comment by arporter** [comment]
Some tasks:
  1. Create a new psyclone driver-script (so that we don't have to invoke PSyclone by doing `python generator.py` anymore)
  2. Decide how we handle our examples when installing from pip (do we ignore them?)
  3. Move the logo into the doc directory?
  4. Copy the top-level `__init__.py` into all sub-directories of the new structure that we want installed
  5. Avoid duplicating the setting of VERSION (in setup.py and docs/conf.py) by e.g. getting setup.py to create a python file/module containing the current version. This can be imported by conf.py.
  6. Update the documentation for the new structure (including new installation instructions)
  7. Update the examples to use the new psyclone driver script.

**Comment by arporter** [comment]
Tests now run and pass in Travis. Coverage report is handled by Coveralls.
Need to update examples and documentation (including the README) to reflect the new world.

**Comment by arporter** [comment]
Have followed what is done in f2py and have added a `write_version_py()` routine to setup.py. This generates a python module containing the current version-number of PSyclone. `doc/conf.py` then
imports this module to get the version number to use when generating the documentation.

**Comment by arporter** [comment]
I've now tackled all 7 points in my TODO list. I think all that remains is to attempt to get PSyclone on to pypi.

**Comment by arporter** [comment]
I'd got the version number wrong in some of my documentation changes. Have now figured out a solution for using the Sphinx ``|release|`` macro within code fragments. Have to do, e.g.:

    .. parsed-literal::
        git checkout tags/\ |release|\ 

This uses escaped-whitespace around the macro name.

Have successfully uploaded to the test pypi server (https://testpypi.python.org/pypi/PSyclone/1.4.1). However, the version of the code that I have re-structured on this branch is *not* 1.4.1 because several issues have been tackled since that release.

**Comment by arporter** [comment]
Have made what I think is the final required change. However, this highlighted that, following the changes on this branch, the `contributions/install` script will not work. @TeranIvy Will the install
script still be required by the Met Office or can you make use of e.g. `python setup.py install`?

**Comment by TeranIvy** [comment]
I will try installing via setup.py tomorrow. Can you point me to the branch? If pip is going to be preferred way of installing PSyclone from now on I will talk with Matthew about getting pip for our 2.7 distribution. The default 2.6 is managed by the MO admins and we can't modify it directly.  Could we use easy_install instead of pip?

**Comment by arporter** [comment]
Hi Iva - the branch is `pypi_restructure`. If easy_install can use setup.py then I think it will work. In fact, if it works for fparser it should work for PSyclone :-)

**Comment by TeranIvy** [comment]
Hi Andy! 
I installed PSyclone from your branch using `python setup.py install` and the full installation schema with specified `bin/` and `lib/` subdirectories, after installing fparser in similar way. This seems a bit complicated, but it keeps PSyclone and fparser separate from our Python 2.7 distribution as we decided before. As the lfric admin I can use `easy_install-<version>` with specific location and package name for anything in Python Package Index, so this will work for PSyclone as well. The only potential problem I can envisage is that `easy_install-2.7 $PSYCLONE_INSTALLDIR psyclone` will put everything in the same directory without separating into `bin/` and `lib/`as it does for fparser as well. However, the full schema works and I will write a bash script for our purposes to automate the process.
To summarise, please remove `contributions/install` :-)

**Comment by arporter** [comment]
Great - will do!

**Comment by TeranIvy** [comment]
P.S. I just realised that `easy_install` has options for `script-dir` so there's the `bin/` problem solved :-)

**Comment by TeranIvy** [comment]
Hello, I have an observation and a question. After installing PSyclone as described above, I get the `$PSYCLONE_INSTALLDIR/bin/psyclone` executable file with the following text:
```
#!/usr/local/sci/bin/python2.7
# EASY-INSTALL-SCRIPT: 'PSyclone==1.4.1','psyclone'
__requires__ = 'PSyclone==1.4.1'
import pkg_resources
pkg_resources.run_script('PSyclone==1.4.1', 'psyclone')
```
How is this supposed to work? Is it supposed to point to `generator.py` script which is now in `$PSYCLONE_INSTALLDIR/lib/PSyclone-1.4.1-py2.7.egg/psyclone` directory? If so, that might be problematic as none of the `*.py` scripts in there has got executable permissions (as `generator.py` currently has with symlink `bin/psyclone` pointing to it). Do you think this could cause problems?

**Comment by arporter** [comment]
Hmm, I think that must be some magic created by setup.py/easy_install - it's not anything I've written. I have changed PSyclone so that we have a top-level `psyclone` driver script and that is installed into bin. There is therefore no need to have a sym-link to generator.py (and in fact, doing so will cause problems). Does that answer your question?

**Comment by TeranIvy** [comment]
I am not sure. Could you please give more details about how the directory structure should look like after installation so I can compare what happens for me? Thank you!

**Comment by arporter** [comment]
TBH, I don't worry about what it should look like - I leave that to pip/easy_install. All I do is tell it I want my psyclone.* modules installed on my PYTHONPATH and the `psyclone` script put in my path. If I do a local install then this means in practice that I get a ~/.local/bin/psyclone and the module files are put in ~/.local/lib//python2.7/site-packages/.

**Comment by TeranIvy** [comment]
The .local location works for me as well if I use 'python setup.py install --user' option. I guess the only way to know for sure whether it works is to test it on LFRic code. I'll try that and let you know the outcome.

**Comment by TeranIvy** [comment]
Righty-ho, LFRic compiles and works normally for installation of PSyclone from `pypi_restructure` branch on my local machine (after adjusting `PATH` and `PYTHONPATH`) and this should hold for other target machines we use (they're all Linux). 
Off to write down all steps for future installations and setup...
Have a good weekend!

**Comment by arporter** [comment]
Excellent! Thanks for testing that.

**Comment by rupertford** [comment]
pull request #51 is now in master. Closing this issue.
